File: clojure/allergies/src/allergies.clj
(ns allergies)

(def ^:private allergic-items {:eggs 1
                               :peanuts 2
                               :shellfish 4
                               :strawberries 8
                               :tomatoes 16
                               :chocolate 32
                               :pollen 64
                               :cats 128})

(defn- allergic? [score v] (> (bit-and score v) 0))

(defn allergies
  "Return allergies list for the given score. e.g. 3 -> [:eggs :shellfish]"
  [score]
  (let [total (apply + (vals allergic-items))
        score (bit-and score total)]
    (for [[k v] allergic-items :when (allergic? score v)] k)))

(defn allergic-to?
  "Return true/false to indicate does the given score allergic to given item."
  [score item]
  (allergic? score (allergic-items item)))

File: javascript/allergies/allergies.js
function Allergies (score) {
    'use strict';

    var allergiesTable = {
        eggs: 1,
        peanuts: 2,
        shellfish: 4,
        strawberries: 8,
        tomatoes: 16,
        chocolate: 32,
        pollen: 64,
        cats: 128
    };

    return {
        list: function () {
            var ret = [];
            var keys = Object.keys(allergiesTable);
            var n = score % 256;
            for (var i = keys.length - 1; i >= 0; i --) {
                if (n - allergiesTable[keys[i]] >= 0) {
                    ret.push(keys[i]);
                    n -= allergiesTable[keys[i]];
                }
            }
            return ret.reverse();
        },

        allergicTo: function (src) {
            var list = this.list();
            return list.indexOf(src) !== -1;
        }
    };
}

module.exports = Allergies;

